source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -343,10 +327,11 @@ FilterDataStatus Filter::onData(ProcessorState& state, Buffer::Instance& data, b
       // Fall through
       break;
     }
-    // Send the chunk on the gRPC stream
-    sendBodyChunk(state, data, ProcessorState::CallbackState::StreamedBodyCallback, end_stream);
-    // Move the data to the queue and optionally raise the watermark.
-    state.enqueueStreamingChunk(data, end_stream, true);
+
+    // Need to first enqueue the data into the chunk queue before sending.
+    auto req = setupBodyChunk(state, data, end_stream);
```

Comment:
We are splitting sendBodyChunk() into two operations: 
1) setupBodyChunk(), which is copy the data into message  2) sendBodyChunk()

The reason is that unit test fuzzer detected a problem with sendBodyChunk, it will call stream_->send(), which will automatically call onReceiveMessage(), then handleBodyResonse(), then dequeueStreamingChunk(). However at that moment, the enqueueStreamingChunk() is not called yet, so, nothing in the queue, and the body response is not handled.  
Thus, the change is to call the enqueueStreamingChunk() before sendBodyChunk().  As enqueueStreamingChunk() also moves the data to the queue buffer. i.e, the data will become empty after enqueueStreamingChunk() here. So, need to call setupBodyChunk() before it to setup the request properly.

Reply:
enqueueStreamingChunk() will move the data in tha consolidate buffer: received_data_
https://github.com/envoyproxy/envoy/blob/21af65095760db381347c031971aa5b3668b8968/source/extensions/filters/http/ext_proc/processor_state.h#L57
There is no easy way to read the last piece out without another move() or copyOut() operation.  The proposed approach has least memory operations.

Reply:
`enqueueStreamingChunk` calls ` chunk_queue_.push`, then it calls `queue_.push_back`, where `queue_` is std::deque<QueuedChunkPtr> queue_.  So, I am not sure I fully understand why we cannot read the data from that deque.

Also, here we read the data from the queue right after the `enqueueStreamingChunk` , which  should not involve consolidate. Did I miss something?

Reply:
The chunk queue only maintain a logical view of the data.

Reply:
could you please elaborate where is the one more copy?

Reply:
ah my bad. I misread your first comment that `setupBodyChunk(), which is copy the data into message` and thought there is an additional copy.

Just took a closer look at code change: basically that copy is to build the request message and already exists in the code. Here you only split the operations

Reply:
Yeah, that's right.